source/win32/low_latency_iocp_context.cpp

```diff
@@ -0,0 +1,682 @@
+        DWORD errorCode = ::GetLastError();
+        throw std::system_error{
+            static_cast<int>(errorCode), std::system_category(),
+            "CreateIoCompletionPort"};
+    }
+
+    const BOOL ok = ::SetFileCompletionNotificationModes(
+        fileHandle,
+        FILE_SKIP_COMPLETION_PORT_ON_SUCCESS |
+        FILE_SKIP_SET_EVENT_ON_HANDLE);
+    if (!ok) {
+        DWORD errorCode = ::GetLastError();
+        throw std::system_error{
+            static_cast<int>(errorCode), std::system_category(),
+            "SetFileCompletionNotificationModes"};
+    }
+}
+
+void low_latency_iocp_context::io_operation::cancel_io() noexcept {
+    assert(ioState != nullptr);
+
+    // Cancel operations in reverse order so that later operations
+    // are cancelled first and don't accidentally end up with earlier
+    // operations being cancelled and later ones completing due to
+    // a race.
+    for (std::uint16_t i = ioState->operationCount; i != 0; --i) {
+        win32::overlapped* o = &ioState->operations[i-1];
+        (void)::CancelIoEx(fileHandle, reinterpret_cast<OVERLAPPED*>(o));
+    }
+}
+
+bool low_latency_iocp_context::io_operation::is_complete() noexcept {
+    assert(context.is_running_on_io_thread());
+    assert(ioState != nullptr);
+
+    if (ioState->pendingCompletionNotifications == 0) {
+        return true;
+    }
+
+    for (std::size_t i = 0; i < ioState->operationCount; ++i) {
+        overlapped* o = &ioState->operations[i];
+
+        // Mark volatile to indicate to the compiler that it might change in the
+        // background. The kernel might update it as the I/O completes.
+        volatile io_status_block* iosb = reinterpret_cast<io_status_block*>(o);
+        if (iosb->Status == STATUS_PENDING) {
+            return false;
+        }
+    }
+
+    // Make sure that we have visibility of the results of all of the I/O
+    // operations.
+    // Assuming that the OS used a "release" memory semantic when writing
+    // to the 'Status' field here.
+    std::atomic_thread_fence(std::memory_order_acquire);
+
+    return true;
+}
+
+bool low_latency_iocp_context::io_operation::start_read(
+        span<std::byte> buffer) noexcept {
+    assert(context.is_running_on_io_thread());
+    assert(ioState != nullptr);
+    assert(ioState->operationCount < max_vectored_io_size);
+
+    bool allowStartingMore = false;
+
+    std::size_t offset = 0;
+    while (offset < buffer.size()) {
+        auto& op = ioState->operations[ioState->operationCount];
+        ++ioState->operationCount;
+
+        op.Internal = 0;
+        op.InternalHigh = 0;
+        op.Offset = 0;
+        op.OffsetHigh = 0;
+        op.hEvent = nullptr;
+
+        // Truncate over-large chunks to a number of bytes that will still
+        // preserve alignment requirements of the underlying device if this
+        // happens to be an unbuffered storage device.
+        // In this case we truncate to the largest multiple of 64k less than
```

Comment:
I think 4Kb would be enough. Any single device will support 4Kb forever. Only RAID arrays have larger fundamental i/o sizes, and that could be anything at all. For example, 128Kb stripe sizes are very common.